engine/src/main/java/org/terasology/engine/subsystem/rpc/DiscordRPCSubSystem.java

```diff
@@ -0,0 +1,308 @@
+        enabled = false;
+        dontTryAgain = true;
+    }
+
+    public void sendRichPresence(RichPresence richPresence) {
+        this.lastRichPresence = richPresence;
+        if (!ready || lastRichPresence == null || !enabled) {
+            return;
+        }
+        ipcClient.sendRichPresence(lastRichPresence);
+    }
+
+    @Override
+    public void onReady(IPCClient client) {
+        if (reconnecting) {
+            getLogger().info("Discord RPC >> Reconnected!");
+            reconnectTries = 1;
+        } else {
+            getLogger().info("Discord RPC >> Connected!");
+            connectedBefore = true;
+        }
+        this.ipcClient = client;
+        if (!ready) {
+            ready = true;
+        }
+        if (lastRichPresence == null) {
+            RichPresence.Builder builder = new RichPresence.Builder();
+            builder.setLargeImage(LARGE_IMAGE);
+            lastRichPresence = builder.build();
+        }
+        client.sendRichPresence(lastRichPresence);
+    }
+
+    @Override
+    public void onDisconnect(IPCClient client, Throwable t) {
+        if (ready) {
+            ready = false;
+        }
+        getLogger().info("Discord RPC >> Disconnected!");
+    }
+
+    @Override
+    public void run() {
+        while (autoReconnect) {
+            try {
+                // Ignore if the Discord RPC is not enabled
+                if (!enabled) {
+                    if (ready) {
+                        try {
+                            getInstance().ipcClient.close();
+                        } catch (Exception ex) { }
+                    }
+                    Thread.sleep(1);
+                    continue;
+                }
+
+                // Don't try to retry to discover the RPC until some reasons happen
+                if (dontTryAgain) {
+                    Thread.sleep(1);
+                    continue;
+                }
+
+                // Connect if the connect on init didn't connect successfully
+                if (!connectedBefore && !ready) {
+                    lastPing = 0;
+                    try {
+                        ipcClient.connect();
+                    } catch (Exception ex) { }
+                    Thread.sleep(15 * 1000);
+                    if (!ready) {
+                        reconnectTries += 1;
+                        if (reconnectTries >= 5) {
+                            dontTryAgain = true;
+                        }
+                    }
+                    continue;
+                }
+
+
+                // Ping to make sure that the RPC is alive
+                if (ready) {
+                    Thread.sleep(1);
+                    lastPing += 1;
+                    if (lastPing >= 5 * 1000) {
+                        ipcClient.sendRichPresence(this.lastRichPresence);
+                        this.lastPing = 0;
+                    }
+                } else { // Reconnect to RPC with tries ( min: 2 seconds, max: 30 seconds )
```

Comment:
This comment is a good example of how nice constants can explain how everything works without even needing a comment :-)

Reply:
**Done** I did too much stuff with the constant variables. This comment is useless I removed it.